### DISCUSSION TITLE
Error handling?

### DISCUSSION THREAD

**Discussion by ttytm** [capy-language/capy#37] [Q&A]
The language looks great. Hats off for the work that goes into this. 

One of the frist things I check when coming across a new language is how it handles errors. Skimming the Readme and examples I'm not finding any references. Can you please point me to the parts I missed?

**Comment by NotAFlyingGoose** [comment]
Thank you! Me and lena are slowly working on improving the language a lot, and it's a *lot* better than it was about a year ago when I made the reddit post to r/programminglanguages. Unfortunately there's still some stuff left before I feel like it's the minimal viable language in which you can get stuff done. We're about 95% of the way there. One of the very last steps we're working on is error handling.

Currently, if there's an error the program will just panic. This isn't ideal. For a solution, I've been very inspired by Rust's enums. The Capy compiler itself uses a lot of enums all over the place, and they're an incredibly useful construct for dealing with varying state.

Some other languages decide to use a syntax for error handling similar to `result, err := frobnicate(walrus);` but in my eyes the problem with this is that the language would have to support null values for this to work (which is a big no), and you aren't *forced* to handle the error. In languages like Rust, you can always call `unwrap`, but that's still a way of handling the error. If you don't even read the value of err, then that might introduce some UB (or ig logical error would be the correct term) where you're messing with null or default values as if there wasn't an error and you have no idea that there was one.

So to sum it up, the language isn't quite perfect yet, and one of the last things we're working on is adding tagged unions with which you can handle errors as values just like any other.

**Comment by ttytm** [comment]
It's visible that some great work is happening 😊 Thanks for the efforts!

Yep, errors as values is imo indispensable for a modern language. How it extends the possibilities for designing APIs of libraries and for failure handling in general, is immense.

I hope I'm not overdoing it by sharing some experiences from the languages Capy leans on.

- When working with Rust, I love the simple error wrapping that `anyhow` provides. Using a result type that wraps the final type `<Result<T>>`. It can be unwrapped, handled in any form or way, and propagated with a simple `.?`. It served (or when it wasn't used, it would have served) the needs for the vast majority of cases I have encountered in Rust projects.

- When it comes to Odin, I think that Ginger Bill is one of the best language designers of the era. However, regarding error handling, I think he is wrong. I can't get behind the statements he makes in `A Reply to The Road to Zig 1.0`
  https://www.gingerbill.org/article/2019/05/13/a-reply-to-the-road-to-zig/#error-handling.
  And while Odin has `or_return` and `or_continue`, one still easily ends up in the Go dilemma of repetitive `if err != nil` checks, because there is no uniform result/error type (which is sometimes better, sometimes worse). For me personally, error handling (and debugging Odin code :D) disqualifies Odin as an everyday use language, which is a pity for an otherwise excellent language. And you already say that null values are not an option.

- Zig's approach with its Result Type `!` (like `fn .. !void`), `try`, `orelse` is convenient. It comes close to an optimum, but from the impression that I get from Capy, it might be too verbose.

<br>

What are your thoughts on having a Result Type `!` `!T` (under the hood this would be a `!<Type, ErrorInterface>)`? Maybe something like this is a fitting combination of Rust, Odin, and Zig.

<details open><summary>Code example</summary>

```c
Person :: struct {
    name: str,
    age: i32
};

birth :: () -> !Person {
    ...
}
```

To access the final type, the result/error must be processed:
```c
person := birth()! // propagate / pass it up the stack.
person := birth() or { Person{} } // assign a fallback default in case of error (must satisfy the final type).
person := birth() or { otherProcReturningPerson() }


// First call then process
birth_res := birth() 
person := birth_res!


// Alternatively ways of unwrapping are return-/loop-breaking-(or similar) statements.
birth() or { return }
birth() or { continue }


// Multiline or-block
birth() or {  
  core.eprintln(err); // access error value
  
  // End an with an according value or a statement like `return`
  return
}
```

Of course, these ideas are not mine; they come up in several other places, so I'm just sharing others' genius.
</details>

In any case, please let me know when you have something concrete @NotAFlyingGoose. I'd like to give the language a closer look and use it then.
